kafka-connect-elastic6/src/main/scala/com/datamountaineer/streamreactor/connect/elastic6/ElasticJsonWriter.scala

```diff
@@ -103,23 +114,37 @@ class ElasticJsonWriter(client: KElasticClient, settings: ElasticSettings)
             .map { batch =>
               val indexes = batch.map { r =>
 
+                val struct = r.newFilteredRecordAsStruct(projections)
+                val payload = simpleJsonConverter.fromConnectData(struct.schema(), struct)
                 val (json, pks) = if (kcqlValue.primaryKeysPath.isEmpty) {
-                  (Transform(
-                    kcqlValue.fields,
-                    kcqlValue.ignoredFields,
-                    r.valueSchema(),
-                    r.value(),
-                    kcql.hasRetainStructure
-                  ), Seq.empty)
+                  (payload, Seq.empty)
                 } else {
-                  TransformAndExtractPK(
-                    kcqlValue.fields,
-                    kcqlValue.ignoredFields,
-                    kcqlValue.primaryKeysPath,
-                    r.valueSchema(),
+                  //extract will flatten if parent/child detected
+                  val keys = kcql.getPrimaryKeys.asScala.map(pk => (pk.toString, pk.toString.replaceAll("\\.", "_"))).toMap
+
+                  val pkValues = Try(r.extract(
                     r.value(),
-                    kcql.hasRetainStructure)
+                    r.valueSchema(),
+                    keys,
+                    Set.empty)) match {
+                    case Success(value) =>
+                      val helper = StructHelper.StructExtension(value)
+                      keys
+                        .values
+                        .map(helper.extractValueFromPath)
+                        .map {
```

Comment:
Would it be better to merge these two map calls so that we're not iterating over the same collection twice?

`     .map(e => helper.extractValueFromPath(e) match {
                          case Right(value) => value.toString
                 // etc etc
`

Reply:
Done